kernel/rtlil.cc

```diff
@@ -4435,8 +4435,39 @@ RTLIL::SigSpec RTLIL::SigSpec::extract(int offset, int length) const
 	log_assert(offset >= 0);
 	log_assert(length >= 0);
 	log_assert(offset + length <= width_);
-	unpack();
+
 	cover("kernel.rtlil.sigspec.extract_pos");
+
+	if (packed())
+	{
+		if (chunks_.size() == 1)
+			return chunks_[0].extract(offset, length);
+
+		SigSpec extracted;
+		int end = offset + length;
+		int chunk_end = 0;
+
+		for (auto const &chunk : chunks_)
+		{
+			int chunk_begin = chunk_end;
+			chunk_end += chunk.width;
+			int extract_begin = std::max(chunk_begin, offset);
+			int extract_end = std::min(chunk_end, end);
+			if (extract_begin >= extract_end)
+				continue;
+			int extract_offset = extract_begin - chunk_begin;
+			int extract_len = extract_end - extract_begin;
+			if (extract_offset == 0 && extract_len == chunk.width)
+				extracted.chunks_.emplace_back(chunk);
+			else
+				extracted.chunks_.emplace_back(
+					chunk.extract(extract_offset, extract_len));
+		}
```

Comment:
I feel like this is more complicated then it needs to be. What about this?

```
SigSpec extracted;
extracted.width_ = length;

auto it = chunks_.begin();
for (; it != chunks_.end() && offset; offset -= it->width, it++) {
    if (offset < it->width) {
        extracted.chunks_.emplace_back(it->extract(offset, it->width - offset));
        length -= it->width - offset;
        break;
    }
}
for (; it != chunks_.end() && length; length -= it->width, it++) {
    if (length >= it->width) {
        extracted.chunks_.emplace_back(*it);
    } else {
        extracted.chunks_.emplace_back(it->extract(0, length));
        break;
    }
}

return extracted;
```
Not that I have compile-tested it.

Reply:
The `it != chunks_.end()` checks should be superfluous even.

Reply:
I don't really care about the specific implementation, the way I wrote it was just the first thing I thought of where I could quickly convince myself of its correctness and when it made this disappear from the profile I was looking at I didn't spend any time thinking of alternatives.

Reply:
Sure, but now that we know the perfomance implication I propose the above.

Reply:
and it's missing an `it++;` above the first break